docs/architecture/psa-driver-interface.md

```diff
@@ -0,0 +1,561 @@
+
+The driver interface allows implementations of the PSA Crypytography API to be built compositionally. An implementation of the PSA Cryptography API is composed of a **core** and zero or more **drivers**. The core handles key management, enforces key usage policies, and dispatches cryptographic operations either to the applicable driver or to built-in code.
+
+Functions in the PSA Cryptography API invoke functions in the core. Code from the core calls drivers as described in the present document.
+
+### Types of drivers
+
+The PSA Cryptography driver interface supports two types of cryptoprocessors, and accordingly two types of drivers.
+
+* **Transparent** drivers implement cryptographic operations on keys that are provided in cleartext at the beginning of each operation. They are typically used for hardware **accelerators**. When a transparent driver is available for a particular combination of parameters (cryptographic algorithm, key type and size, etc.), it is used instead of the default software implementation. Transparent drivers can also be pure software implementations that are distributed as plug-ins to a PSA Crypto implementation.
+* **Opaque** drivers implement cryptographic operations on keys that can only be used inside a protected environment such as a **secure element**, a hardware security module, a smartcard, a secure enclave, etc. An opaque driver is invoked for the specific key location that the driver is registered for: the dispatch is based on the key's lifetime.
+
+### Requirements
+
+The present specification was designed to fulfil the following high-level requirements.
+
+[Req.plugins] It is possible to combine multiple drivers from different providers into the same implementation, without any prior arrangement other than choosing certain names and values from disjoint namespaces.
+
+[Req.compile] It is possible to compile the code of each driver and of the core separately, and link them together. A small amount of glue code may need to be compiled once the list of drivers is available.
+
+[Req.types] Support drivers for the following types of hardware: accelerators that operate on keys in cleartext; cryptoprocessors that can wrap keys with a built-in keys but not store user keys; and cryptoprocessors that store key material.
+
+[Req.portable] The interface between drivers and the core does not involve any platform-specific consideration. Driver calls are simple C functions. Interactions between driver code and hardware happen inside the driver (and in fact a driver need not involve any hardware at all).
+
+[Req.location] Applications can tell which location values correspond to which secure element drivers.
+
+[Req.fallback] Accelerator drivers can specify that they do not fully support a cryptographic mechanism and that a fallback to core code may be necessary. Conversely, if an accelerator fully supports cryptographic mechanism, the core does not need to include code for this mechanism.
```

Comment:
Can this be achieved through re-implementing the core implementations as standalone software-backed accelerator drivers? It might lead to pulverisation of the core code, though, depending on how granular (see other comment) an accelerator can define what it can or cannot support.

Reply:
Then there would be only fallback from one driver to another driver, not the last fallback to built-in code, but that wouldn't really change the fallback mechanism.

Reply:
Well, the last fallback to built-in code would then just be fallback to the last 'driver' (which is the built-in code). If that is wrapped in e.g. a `#if (FALLBACK_ENABLED)`, followed by similar nested clauses per granularity of capabilites, wouldn't that make it easier to compile in-out software fallback as needed? And have the linker drop core implementations when no references to it are made due to the compiling out?

Reply:
Additional question: in which order are accelerator drivers called if there are multiple accelerators with overlapping capabilities?

Reply:
> wouldn't that make it easier to compile in-out software fallback as needed?

Yes, but that means the glue code generator can't run until the application configuration is available. There's some tension here between core implementer convenience (a single mechanism to define what gets included) and platform developer and application developer convenience (define their aspects of the configuration independently).

Reply:
Regarding the order of accelerator drivers: see the “Fallback” section.

> If there are multiple available transparent drivers, the core tries them in turn until one is declared without a true `"fallback"` property or returns a status other than `PSA_ERROR_NOT_SUPPORTED`.